### PR TITLE
fix beam search script

### PR DESCRIPTION
**Pull request by szha**
## Description ##
The following run was broken.
`python beam_search_generator.py --bos I think --lm standard_lstm_lm_200`

This PR fixes the issue by adding state_info to language model while preserving existing behaviors.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] add state_info to models
- [x] use state_info to decide state shape in beam search

### CODE REVIEW

**Comment by szha** [comment]
@sxjscience @hhexiy

**Comment by mli** [comment]
Job PR-175/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-175/6/index.html

File: gluonnlp/model/beam_search.py [hunk 1]
```diff
@@ -92,30 +92,56 @@ def _expand_to_beam_size(data, beam_size, batch_size):
         Each NDArray should have shape (batch_size * beam_size, ...)
     """
     if isinstance(data, list):
-        return [_expand_to_beam_size(ele, beam_size, batch_size) for ele in data]
+        assert not state_info or isinstance(state_info, list), \
```
Comment (szhengac) [inline gluonnlp/model/beam_search.py]: I think we can move this part outside if else by using 
`assert not state_info or isinstance(state_info, type(data))`

File: gluonnlp/model/beam_search.py [hunk 2]
```diff
@@ -124,14 +150,37 @@ def _choose_states(F, states, indices):
         Each NDArray/Symbol should have shape (N, ...).
     """
     if isinstance(states, list):
-        return [_choose_states(F, ele, indices) for ele in states]
+        assert not state_info or isinstance(state_info, list), \
```
Comment (szhengac) [inline gluonnlp/model/beam_search.py]: Same as previous comment

File: gluonnlp/model/beam_search.py [hunk 3]
```diff
@@ -92,30 +92,56 @@ def _expand_to_beam_size(data, beam_size, batch_size):
...
 
-def _choose_states(F, states, indices):
+def _choose_states(F, states, state_info, indices):
     """
 
     Parameters
     ----------
     F : ndarray or symbol
     states : Object contains NDArrays/Symbols
         Each NDArray/Symbol should have shape (N, ...).
```
Comment (szhengac) [inline gluonnlp/model/beam_search.py]: Fix docstring. Now `N` may not be in the first dim.

File: gluonnlp/model/beam_search.py [hunk 4]
```diff
@@ -124,14 +150,37 @@ def _choose_states(F, states, indices):
...
+                'state and state_info doesn\'t match'
+        if not state_info:
+            state_info = {'__layout__': 'NC'}
+        batch_axis = state_info['__layout__'].find('N')
+        if batch_axis != 0:
+            states = states.swapaxes(0, batch_axis)
+        states = F.take(states, indices)
+        if batch_axis != 0:
+            states = states.swapaxes(0, batch_axis)
+        return states
```
Comment (szhengac) [inline gluonnlp/model/beam_search.py]: So the returned states always have batch in the first dim even it is not the case for the input? Would it cause some inconsistency?

Comment (szha) [reply gluonnlp/model/beam_search.py]: Because of the two swaps, the batch dimension should be where it used to be.

**Comment by szhengac** [comment]
Also, test case with different layouts is needed.

File: gluonnlp/model/beam_search.py (hunk 4 above)
Comment (szhengac) [reply gluonnlp/model/beam_search.py]: Yes, you are right.

File: gluonnlp/model/beam_search.py [hunk 5]
```diff
@@ -75,73 +75,118 @@ def hybrid_forward(self, F, log_probs, scores, step):   # pylint: disable=argume
...
     elif isinstance(data, mx.nd.NDArray):
-        if data.shape[0] != batch_size:
-            raise ValueError('The leading dimension of all the inner elements in states must be '
-                             '{}, Find shape={}'.format(batch_size, data.shape))
-        return data.reshape((batch_size, 1) + data.shape[1:])\
-                   .broadcast_axes(axis=1, size=beam_size)\
-                   .reshape((batch_size * beam_size,) + data.shape[1:])
+        if not state_info:
+            state_info = {'__layout__': 'NC'}
+        batch_axis = state_info['__layout__'].find('N')
```
Comment (sxjscience) [inline gluonnlp/model/beam_search.py]: I think the following will be better:
```python
if not state_info:
   batch_axis = 0
else:
   batch_axis = state_info['__layout__'].find('N')
```

File: gluonnlp/model/beam_search.py (hunk 4 above)
Comment (sxjscience) [reply gluonnlp/model/beam_search.py]: We can directly use take(axis=batch_axis) once https://github.com/apache/incubator-mxnet/pull/11326 is merged.

File: tests/unittest/test_beam_search.py [hunk 6]
```diff
@@ -28,44 +28,82 @@ def test_beam_search_score():
...
             updated_states = []
-            for beam_id, use_new in zip(sel_beam_ids, use_new_states):
-                if use_new:
-                    updated_states.append(new_states[beam_id])
-                else:
-                    updated_states.append(states[beam_id])
-            return mx.nd.stack(*updated_states, axis=0)
+            if not state_info:
+                state_info = {'__layout__': 'NC'}
+            batch_axis = state_info['__layout__'].find('N')
```
Comment (sxjscience) [inline tests/unittest/test_beam_search.py]: Same here
```python
if not state_info:
  batch_axis = 0
else:
 ...
```

Comment (sxjscience) [reply tests/unittest/test_beam_search.py]: Since it's in the test, the change is not necessary.

**Comment by mli** [comment]
Job PR-175/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-175/13/index.html

**Comment by mli** [comment]
Job PR-175/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-175/18/index.html

**Pull request merged**